### PR TITLE
RequestHeader.path() is unsafe

### PR DESCRIPTION
**Pull request by jroper**
I have Play 2.3.6 Java app. In it I have my own error handler for handling bad requests, method looks like:

    public Promise<play.mvc.Result> onBadRequest(RequestHeader request, String error {
        Logger.debug("400 error: " + error + ", " + request.path() + ", " + request.remoteAddress());
        return Promise.<play.mvc.Result> pure(badRequest(views.html.error404.render()));
    }
When an error occurs, instead of my own error page I am getting standard error page. It turns out problem is in request.path() method which is supposed to return string. But if the URL is incorrect it throws an exception and breaks my error handling code. 

An example of bad request causing it looks like this: https://www.playframework.com/documentation/2.3.x/Home%
(with % at the end of url).

### CODE REVIEW

**Comment by richdougherty** [comment]
A couple of errors. The `AkkaHttpFlashCookieSpec` is probably fixed now that Akka HTTP has been upgraded. @play-pull-request-validator, please run again to check this.

I'm not sure what the problem is with `NettyBadClientHandlingSpec`...

```
[info] NettyBadClientHandlingSpec
[info] 
[info] Play should
[info] ! gracefully handle long urls and return 414
[error]    RuntimeException: : Exception while reading response 1 java.net.SocketException: Connection reset  (BasicHttpClient.scala:203)
[error] play.it.http.BasicHttpClient.readResponse(BasicHttpClient.scala:203)
[error] play.it.http.BasicHttpClient$$anonfun$sendRequest$4.apply(BasicHttpClient.scala:119)
[error] play.it.http.BasicHttpClient$$anonfun$sendRequest$4.apply(BasicHttpClient.scala:117)
[error] play.it.http.BasicHttpClient.sendRequest(BasicHttpClient.scala:117)
[error] play.it.http.BasicHttpClient$$anonfun$1.apply(BasicHttpClient.scala:31)
[error] play.it.http.BasicHttpClient$$anonfun$1.apply(BasicHttpClient.scala:29)
[error] play.it.http.BasicHttpClient$.makeRequests(BasicHttpClient.scala:29)
...
```

```
[ERROR] [12/29/2014 01:54:36.699] [play-akka-http-server-akka.actor.default-dispatcher-9] [ActorSystem(play-akka-http-server)] Internal server error, sending 500 response
java.lang.RuntimeException: Error parsing response headers: List(ErrorInfo(Illegal HTTP header 'Set-Cookie',7))
	at scala.sys.package$.error(package.scala:27)
	at play.core.server.akkahttp.ModelConversion$.convertResponseHeaders(ModelConversion.scala:219)
	at play.core.server.akkahttp.ModelConversion$.convertResult(ModelConversion.scala:118)
	at play.core.server.akkahttp.AkkaHttpServer$$anonfun$7.apply(AkkaHttpServer.scala:156)
	at play.core.server.akkahttp.AkkaHttpServer$$anonfun$7.apply(AkkaHttpServer.scala:154)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:46)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
	at scala.concurrent.Promise$class.complete(Promise.scala:55)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:23)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[info] AkkaHttpFlashCookieSpec
[info] 
[info] the flash cookie should
[info] + can be set for one request
[info] x be removed after a redirect
[error]    '500' is not equal to '200' (FlashCookieSpec.scala:51)
```

**Comment by jroper** [comment]
Hmm...

**Comment by richdougherty** [comment]
I think this is a Java IO or Commons IO bug that I solved once before. Unfortunately I can't remember exactly what the problem or solution was. I think what I ended up doing was reading the input bytes myself, one byte at a time, so that I could handle EOF properly.

**Comment by richdougherty** [comment]
Ah, found it: https://github.com/playframework/prune/blob/134e6812c5dfae5072deea79563f8efe0c5d6cdc/src/main/scala/com/typesafe/play/prune/Exec.scala#L177-L195

**Comment by jroper** [comment]
@play-pull-request-validator try again.

**Comment by richdougherty** [comment]
See also discussion about Akka HTTP parse error handling: https://github.com/akka/akka/issues/16606

**Comment by benmccann** [comment]
@play-pull-request-validator try again

**Comment by benmccann** [comment]
@szymond the test errors look like they're due to this PR

**Comment by jroper** [comment]
@benmccann the PRs mine :) I haven't got around to working out what's going wrong - it seems to be some sort of bug in the JDK.

**Comment by jroper** [comment]
@play-pull-request-validator

**Comment by benmccann** [comment]
@play-pull-request-validator try again. does it work any better on java 8? :-)

**Comment by jroper** [comment]
It seems to be a problem when you try to read from a connection where you haven't encountered EOF yet while reading, but the server has closed the connection.  Simple work around - I've just made us send a content length of 0 with the request uri too long response, this way we can simply not make any reads, and it's better anyway because it's always better to send a content length than to force the client to read and find the connection has closed.

File: framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala [hunk 1]
```diff
@@ -18,30 +18,52 @@ trait BadClientHandlingSpec extends PlaySpecification with ServerIntegrationSpec
...
+        Future.successful(Results.BadRequest("Bad path: " + request.path))
+      def onServerError(request: RequestHeader, exception: Throwable) = Future.successful(Results.Ok)
+    }) { port =>
+      val response = BasicHttpClient.makeRequests(port)(
+        BasicRequest("GET", "/[", "HTTP/1.1", Map(), "")
+      )(0)
+
+      response.status must_== 400
+      response.body must beLeft("Bad path: /[")
+    }.skipUntilAkkaHttpFixed
```
Comment (richdougherty) [inline framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala]: Is this something we plan to eventually fix in the Akka HTTP backend?

File: framework/src/play/src/main/scala/play/api/Routes.scala [hunk 2]
```diff
@@ -1,81 +1,110 @@
...
+      host,
+      host,
+      routes.map { route =>
+        "_nS('%s'); _root.%s = %s".format(
+          route.name.split('.').dropRight(1).mkString("."),
+          route.name.split('.').map(name => if (jsReservedWords.contains(name)) { "['" + name + "']" } else { "." + name }).mkString("").tail,
+          route.f)
+      }.mkString("\n"),
+      name)
   }
```
Comment (richdougherty) [inline framework/src/play/src/main/scala/play/api/Routes.scala]: I assume there were no changes in the JS stuff, it's just an indentation change?

File: framework/src/play-netty-server/src/main/scala/play/core/server/netty/PlayDefaultUpstreamHandler.scala [hunk 3]
```diff
@@ -120,14 +127,10 @@ private[play] class PlayDefaultUpstreamHandler(server: Server, allChannels: Defa
...
-            val rh = tryToCreateRequest
-            // Force parsing of uri
-            rh.path
-            rh
-          }.fold(
+          .allCatch[RequestHeader].either(tryToCreateRequest).fold(
             e => {
-              val rh = createRequestHeader()
+              // use unparsed path
+              val rh = createRequestHeader(nettyUri.getPath)
```
Comment (richdougherty) [inline framework/src/play-netty-server/src/main/scala/play/core/server/netty/PlayDefaultUpstreamHandler.scala]: The core logic change LGTM.

File: framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala (hunk 1 above)
Comment (jroper) [inline framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala]: I don't know.

File: framework/src/play/src/main/scala/play/api/Routes.scala (hunk 2 above)
Comment (jroper) [inline framework/src/play/src/main/scala/play/api/Routes.scala]: Yes.  There used to be curly braces around the whole package, I removed them.

File: framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala (hunk 1 above)
Comment (mkurz) [reply framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala:66]: Greetings from 2023. Fixed in #11671 finally :wink:

Comment (jroper) [reply framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala:66]: Thankyou!

**Pull request merged**